source/validate.h

```diff
@@ -84,91 +103,303 @@ enum class FunctionDecl {
+  std::vector<BasicBlock*>& get_blocks();
+
+  /// Returns a list of all the cfg constructs in the function
+  const std::list<CFConstruct>& get_constructs() const;
+
+  /// Returns a list of all the cfg constructs in the function
+  std::list<CFConstruct>& get_constructs();
 
   // Returns the number of blocks in the current function being parsed
   size_t get_block_count() const;
 
-  // Returns true if called after a function instruction but before the
-  // function end instruction
-  bool in_function_body() const;
+  /// Returns the id of the funciton
+  uint32_t get_id() const { return id_; };
 
-  // Returns true if called after a label instruction but before a branch
-  // instruction
+  // Returns the number of blocks in the current function being parsed
+  size_t get_undefined_block_count() const;
+  const std::unordered_set<uint32_t>& get_undefined_blocks() const {
+    return undefined_blocks_;
+  }
+
+  /// Returns true if called after a label instruction but before a branch
+  /// instruction
   bool in_block() const;
 
-  libspirv::DiagnosticStream diag(spv_result_t error_code) const;
+  /// Returns the block that is currently being parsed in the binary
+  BasicBlock& get_current_block();
+
+  /// Returns the block that is currently being parsed in the binary
+  const BasicBlock& get_current_block() const;
+
+  /// Prints a GraphViz digraph of the CFG of the current funciton
+  void printDotGraph() const;
+
+  /// Prints a directed graph of the CFG of the current funciton
+  void printBlocks() const;
 
  private:
-  // Parent module
+  /// Parent module
   ValidationState_t& module_;
 
-  // Function IDs in a module
-  std::vector<uint32_t> id_;
+  /// The result id of the OpLabel that defined this block
+  uint32_t id_;
 
-  // OpTypeFunction IDs of each of the id_ functions
-  std::vector<uint32_t> type_id_;
+  /// The type of the function
+  uint32_t function_type_id_;
 
-  // The type of declaration of each function
-  std::vector<FunctionDecl> declaration_type_;
+  /// The type of the return value
+  uint32_t result_type_id_;
 
-  // TODO(umar): Probably needs better abstractions
-  // The beginning of the block of functions
-  std::vector<std::vector<uint32_t>> block_ids_;
+  /// The control fo the funciton
+  SpvFunctionControlMask function_control_;
 
-  // The variable IDs of the functions
-  std::vector<std::vector<uint32_t>> variable_ids_;
+  /// The type of declaration of each function
+  FunctionDecl declaration_type_;
 
-  // The function parameter ids of the functions
-  std::vector<std::vector<uint32_t>> parameter_ids_;
+  /// The beginning of the block of functions
+  std::unordered_map<uint32_t, BasicBlock> blocks_;
 
-  // NOTE: See correspoding getter functions
-  bool in_function_;
-  bool in_block_;
+  /// A list of blocks in the order they appeared in the binary
+  std::vector<BasicBlock*> ordered_blocks_;
+
+  /// Blocks which are forward referenced by blocks but not defined
```

Comment:
Unless this is always up-to-date, I'd keep it out of the class and with the code that actually calculates it.